### ISSUE TITLE
[CUDA] [M5] Phase 3.1: FP16 mixed precision with autocast

### ISSUE THREAD

**Issue by atfrank** [atfrank/MLFF-Distiller#28]
## Objective
Implement FP16 mixed precision using torch.cuda.amp.autocast() to achieve 1.5-2x speedup on RTX 3080 Ti tensor cores.

## Context
- **Current Implementation**: Has bugs (dtype mismatches)
- **Fix Required**: Use autocast() only, NO .half() on model
- **Target**: 1.5-2x speedup with <10 meV accuracy degradation

## Tasks
- [ ] Remove any explicit `.half()` calls on model
- [ ] Implement autocast context manager:
  ```python
  with torch.cuda.amp.autocast(dtype=torch.float16):
      energy = model(...)
      forces = -torch.autograd.grad(energy, positions)[0]
  ```
- [ ] Test on validation set (100 molecules)
- [ ] Measure accuracy degradation:
  - [ ] Energy MAE increase
  - [ ] Force MAE increase
  - [ ] Max energy error
  - [ ] Max force error
- [ ] Benchmark speedup on 10-200 atom molecules
- [ ] Document results in benchmarks/phase3_1_fp16_results.md

## Acceptance Criteria
- [ ] No dtype mismatch errors
- [ ] Speedup ≥1.4x on tensor core operations (matmul, bmm)
- [ ] Energy MAE increase <10 meV
- [ ] Force MAE increase <0.01 eV/Å
- [ ] No NaN or inf values in outputs
- [ ] Works correctly with force computation (autograd)

## Risks
- **Accuracy degradation**: FP16 has less precision than FP32
- **Numerical instability**: Some ops may overflow/underflow in FP16
- **Mixed precision complexity**: Autocast may not cover all ops

## Mitigation
- Validate carefully on diverse test set
- If accuracy loss >1%, use selective mixed precision (some ops in FP32)
- Consider BF16 if available (better dynamic range than FP16)
- Monitor for NaN/inf during training/inference

## Files to Modify
- `src/mlff_distiller/inference/optimized_inference.py`
- `src/mlff_distiller/inference/ase_calculator_optimized.py` (create new)

## Timeline
1 day

## Dependencies
- Depends on #27 (torch.compile()) - should combine both optimizations

## Related Issues
- Part of Phase 3.1 execution plan
